src/components/Idler.js

```diff
@@ -0,0 +1,73 @@
+// @flow
+
+import { PureComponent } from 'react'
+import { createStructuredSelector } from 'reselect'
+import { connect } from 'react-redux'
+import { hasPasswordSelector, autoLockTimeoutSelector } from 'reducers/settings'
+import debounce from 'lodash/debounce'
+import { lock } from 'reducers/application'
+
+type Props = {
+  autoLockTimeout: number,
+  hasPassword: boolean,
+  lock: Function,
+}
+
+type State = {
+  lastAction: number,
+}
+
+const mapStateToProps = createStructuredSelector({
+  autoLockTimeout: autoLockTimeoutSelector,
+  hasPassword: hasPasswordSelector,
+})
+
+const mapDispatchToProps = {
+  lock,
+}
+
+class Idler extends PureComponent<Props, State> {
+  state = {
+    lastAction: -1,
+  }
+
+  componentDidMount() {
+    window.addEventListener('keydown', this.debounceOnChange)
+    window.addEventListener('mouseover', this.debounceOnChange)
+    this.interval = setInterval(this.checkForAutoLock, 10000)
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('keydown', this.debounceOnChange)
+    window.removeEventListener('mouseover', this.debounceOnChange)
+    clearInterval(this.interval)
+  }
+
+  interval: IntervalID
+
+  debounceOnChange = debounce(_ => this.idleTimeHandler(), 1000)
+
+  checkForAutoLock = _ => {
+    const timeout = this.props.autoLockTimeout
+    if (this.props.hasPassword && timeout && +timeout !== -1) {
+      if (Date.now() - (this.state.lastAction + timeout * 60000) > 0) {
+        this.props.lock()
+      }
+    }
+  }
+
+  idleTimeHandler = _ => {
+    this.setState(_ => ({
+      lastAction: Date.now(),
```

Comment:
technically it wouldn't need to be a state. it could literally be a property living on the comopnent instance. but i think it's ok as this render() don't do much anyway